rapids-cmake/cpm/detail/pinning_write_file.cmake

```diff
@@ -0,0 +1,206 @@
+function(rapids_cpm_pinning_add_json_entry json_var package_name url_var sha_var)
+  include("${rapids-cmake-dir}/cpm/detail/get_default_json.cmake")
+  include("${rapids-cmake-dir}/cpm/detail/get_override_json.cmake")
+  get_default_json(${package_name} json_data)
+  get_override_json(${package_name} override_json_data)
+
+  set(url_string)
+  set(sha_string)
+  if(${url_var})
+    string(CONFIGURE [=["git_url": "${${url_var}}",]=] url_string)
+  endif()
+  if(${sha_var})
+    string(CONFIGURE [=["git_tag": "${${sha_var}}",]=] sha_string)
+  endif()
+  # We start with a default template, and only add members that don't exist
+  string(CONFIGURE [=[{
+  "version": "${CPM_PACKAGE_${package_name}_VERSION}",
+  ${url_string}
+  ${sha_string}
+  "git_shallow": false,
+  "always_download": true
+  }]=]
+                   pinned_json_entry)
+
+  foreach(data IN LISTS override_json_data json_data)
+    if(NOT data)
+      # Need to handle both json_data and the override being empty
+      continue()
+    endif()
+    string(JSON entry_count LENGTH "${data}")
+    math(EXPR entry_count "${entry_count} - 1")
+    # cmake-lint: disable=E1120
+    foreach(index RANGE ${entry_count})
+      string(JSON member MEMBER "${data}" ${index})
+      string(JSON existing_value ERROR_VARIABLE dont_have GET "${pinned_json_entry}" ${member})
+      if(dont_have)
+        string(JSON value GET "${data}" ${member})
+        rapids_cpm_pinning_create_and_set_member(pinned_json_entry ${member} ${value})
+      endif()
+    endforeach()
+  endforeach()
+  set(${json_var} "\"${package_name}\": ${pinned_json_entry}" PARENT_SCOPE)
+endfunction()
+
+#[=======================================================================[.rst:
+rapids_cpm_pinning_write_file
+-----------------------------
+
+.. versionadded:: v24.04.00
+
+This function will write out the pinned version info to the provided files when we are in the root
+CMakeLists.txt
+#]=======================================================================]
+function(rapids_cpm_pinning_write_file)
+
+  find_package(Git QUIET REQUIRED)
+
+  set(_rapids_json
+      [=[
+{
+"root": {
+"packages": {
+]=])
+
+  list(POP_BACK CPM_PACKAGES last_package)
+  foreach(package IN LISTS CPM_PACKAGES last_package)
+    # Clear variables so we don't re-use them between packages when one package doesn't have a git
+    # url or sha
+    set(git_url)
+    set(git_sha)
+    set(not_last_package TRUE)
+    if(package STREQUAL last_package)
+      set(not_last_package FALSE)
+    endif()
+    rapids_cpm_pinning_extract_source_git_info(${package} git_url git_sha)
+    rapids_cpm_pinning_add_json_entry(_rapids_entry ${package} git_url git_sha)
+    if(not_last_package)
+      string(APPEND _rapids_entry [=[,
+]=])
+    else()
+      string(APPEND _rapids_entry [=[
+]=])
+    endif()
+    string(APPEND _rapids_json "${_rapids_entry}")
+  endforeach()
+
+  set(post_amble [=[
```

Comment:
Maybe `epilogue` so that we can use a real word :)

Reply:
Just going to drop the variable entirely